File: DynamicProgramming/85_MaximalRectangle.java
public class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix.length==0 || matrix[0].length==0) {
            return 0;
        }

        int colNum = matrix[0].length;
        int[] hight = new int[colNum];
        int[] left = new int[colNum];
        int[] right = new int[colNum];
        for (int i=0; i<colNum; i++) {
            right[i] = colNum;
        }

        int maxV = 0;
        for (int i=0; i<matrix.length; i++) {
            // heigh
            for (int j=0; j<colNum; j++) {
                if (matrix[i][j] == '1') {
                    hight[j] += 1;
                } else {
                    hight[j] = 0;
                }
            }

            // left
            int curr = 0;
            for (int j=0; j<colNum; j++) {
                if (matrix[i][j] == '1') {
                    left[j] = Math.max(left[j], curr);
                } else {
                    left[j] = 0;
                    curr = j+1;
                }
            }

            // right
            curr = colNum;
            for (int j=colNum-1; j>=0; j--) {
                if (matrix[i][j] == '1') {
                    right[j] = Math.min(right[j], curr);
                } else {
                    right[j] = colNum;
                    curr = j;
                }
            }

            for (int j=colNum-1; j>=0; j--) {
                maxV = Math.max((right[j] - left[j]) * hight[j], maxV);
            }
        }
        return maxV;
    }
}

File: DynamicProgramming/85_MaximalRectangle.py
class Solution(object):

    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return 0
        col_num = len(matrix[0])
        hight = [0 for x in range(col_num)]
        left = [0 for x in range(col_num)]
        right = [col_num for x in range(col_num)]
        max_v = 0
        for i in range(len(matrix)):
            for j in range(col_num):
                if matrix[i][j] == "1":
                    hight[j] += 1
                else:
                    hight[j] =0
                
            # calculate left
            curr_left = 0
            for j in range(col_num):
                if matrix[i][j] == "1":
                    left[j] = max(curr_left, left[j])
                else:
                    left[j] = 0
                    curr_left = j + 1

            # calculate right
            curr_right = col_num
            for j in range(col_num-1, -1, -1):
                if matrix[i][j] == "1":
                    right[j] = min(curr_right, right[j])
                else:
                    right[j] = col_num
                    curr_right = j
            
            # max_v = 0
            for j in range(col_num):
                max_v = max((right[j] - left[j]) * hight[j], max_v)

        return max_v
